docs/manual.md

```diff
@@ -807,91 +805,294 @@ user:
+    matches:
+      - 'group1'
+      - 'group2'
+
+  # This fails, since the returned result and it's no longer a subset
+  basic_array_failing:
+    content:
+      - 'group1'
+      - 'group2'
+      - 'group3'
+    matches:
+      - 'group1'
+      - 'group2'
+      - 'group2' # this 2nd group2 is not in the returned content
+```
+
+#### io.Readers
+
+This is the most magical matcher for goss. It remains a default for historic and performance reasons. Some attributes return an io.Reader that is read line by line (ex. file content, command, http body). This allows goss to validate large files/content efficiently.
+
 
-* `"string"` - checks if any line contain string.
-* `"!string"` - inverse of above, checks that no line contains string
-* `"\\!string"` - escape sequence, check if any line contains `"!string"`
-* `"/regex/"` - verifies that line contains regex
-* `"!/regex/"` - inverse of above, checks that no line contains regex
+Each pattern is checked against the attribute output, the type of patterns are:
 
-**NOTE:** Pattern attributes do not support [Advanced Matchers](#advanced-matchers)
+* `"foo"` - checks if any line contains `foo`
+* `"!foo"` - inverse of above, checks that no line contains `foo`
+* `"\\!foo"` - escape sequence, check if any line contains `!string`
+* `"/[Rr]egex/"` - verifies that line matches regex
+* `"!/[Rr]egex/"` - inverse of above, checks that no line matches regex
 
-**NOTE:** Regex support is based on golang's regex engine documented [here](https://golang.org/pkg/regexp/syntax/)
+**NOTE:** Regex support is based on Golang's regex engine documented [here](https://golang.org/pkg/regexp/syntax/)
 
 **NOTE:** You will **need** the double backslash (`\\`) escape for Regex special entities, for example `\\s` for blank spaces.
 
-### Example
-```bash
-$ cat /tmp/test.txt
-found
-!alsofound
+Example:
+
+```yaml
+file:
+  /tmp/test.txt:
+    exists: true
+    contains:
+    - "foo"
+    - "!bar"
+    - "/[Gg]oss/"
+```
 
+The above is roughly similar to:
 
-$ cat goss.yaml
+```yaml
 file:
   /tmp/test.txt:
     exists: true
     contains:
-    - found
-    - /fou.d/
-    - "\\!alsofound"
-    - "!missing"
-    - "!/mis.ing/"
+      and:
+        - contain-element: "foo"
+        - not: {contain-element: "bar"}
+        - contain-element: {match-regexp: "[Gg]oss"}
+
+```
+
+### Transforms
+
+If the system state type and the expected type don't match, goss will attempt to transform the system state type before matching it.
+
+For example, kernel-param attribute returns a string, however, it can be tested using numeric comparisons:
+
+Example kernel-param test:
+```yaml
+kernel-param:
+  net.core.somaxconn:
+      value: "128"
+```
+
+Kernel-param (failing) test with transform:
+```yaml
+kernel-param:
+  net.core.somaxconn:
+      value: {gt: 200}
+```
+
+When a transformed test fails, it will detail the transformers used, the `-o include_raw` option can be used to include the raw, untransformed attribute value:
```

Comment:
This means the tests would need to be re-run to get that extra information. As a user, I would prefer that if a test fails, all of the possible debugging information were given to me so I can have it up-front. Not doing that
* in the best case, costs me the time to re-run the tests
    * but this is harder on machines that are remote
* in the worst case, costs me a lot of time to re-run the tests multiple times in the event of chasing an intermittent failure (much more likely in machine state testing, and more likely again if the test touches an external dependency like `addr`, `http` etc).